### DISCUSSION TITLE
non-linear 2D geometry

### DISCUSSION THREAD

**Discussion by cecidip** [gimli-org/pyGIMLi#413] [Q&A]
Hello, which would be the best way to perform an inversion for a non-linear 2D line, like this: 
![image](https://user-images.githubusercontent.com/22650615/179525023-352938c5-b010-45b8-9f7e-f4066c27d2cd.png)
(in orange the sources, in blue the receivers) .
I processed it as a 2D but on the turns the results are not very good, of course. I just would like to have your opinion on the best approach, if there is some option on the code where i can create a 2D grid but takin  into account its geometry.

Thanks,

**Comment by halbmy** [comment]
To me, it seems like the perfect solution is a hybrid 2D/3D inversion, i.e. forward calculation on a 3D mesh and inversion on a 2D mesh with constant markers in the y direction. I will create an example for it.

Another approach would be to correct the data by computing the geometry effect (in ERT the ratio between 2D and 3D geometric factors, or in seismics the ratio of ray lengths) and correct the data, invert it and project it back to 3D.

**Reply by cecidip** [reply]
Thank you ! I'd appreciate if you could show me an example too.

**Comment by halbmy** [comment]
I have one for ERT (see BERT pdf tutorial Appendix E.3)

![image](https://user-images.githubusercontent.com/7136401/181305192-f1d6be9e-2cd8-4ec2-a4f2-95ebb995cb9d.png)

but none for traveltime. If you can provide your data, this could be a nice example.

**Reply by benbpete** [reply]
Jumping in on this thread.  I have similar resistivity data that is collected along non-linear lines.  I started working through the example above from the BERT manual, and ran into an issue with the harmfit function that I can't seem to resolve. Below is the code I am using, which is basically what is shown in the manual with some minor changes. prior to this snippet I've already imported the appropriate packages sans the harmfit:

```
import pygimli.frameworks.harmfit as hf

tt,yy,xx,hh = np.genfromtxt(path + coordfile, unpack=True, delimiter = ",", skip_header=1,usecols = (0,1,2,3))
#tape distance, northing, easting, elevation

x = hf(xx,tt,nc=10)[0]
y = hf(yy,tt,nc=10)[0]
z = hf(hh,tt,nc=20)[0]

t = np.hstack((0,np.cumsum(np.sqrt(np.diff(x)**2+np.diff(y)**2))))

elevfig,elevax = plt.subplots()
elevax.plot(tt,hh,"bx-");
elevax.plot(tt,z,'r-');
plt.savefig(file + '_topofit.png')

coordfig,coordax = plt.subplots()
coordax.plot(yy,xx,"bx-"); #northing and easting on plot flipped for viewing purposes
#coordax.plot(y,x,"r-");
plt.savefig(file + '_coordfit.png')
```
When I run the code, I end up with a decent looking fit for the elevation data, but it seems like the harmfit function is fitting well beyond the limits of the coordinates.  
![LFA30_topofit](https://user-images.githubusercontent.com/105517351/186889799-37a17a04-96a7-4a23-9b9e-29c6cbde5cd4.png)
![LFA30_coordfit](https://user-images.githubusercontent.com/105517351/186889825-9808ffcb-44e3-4507-87f1-3af597ac44cd.png)

Here are just the coordinates plotted:
![LFA30_coords](https://user-images.githubusercontent.com/105517351/186890113-d43c2246-7e81-44e7-9db5-eea0c0755ee0.png)

and the coordinate file:
[LFA30_(1750M).csv](https://github.com/gimli-org/gimli/files/9432576/LFA30_.1750M.csv)

Any thoughts on what is going on would be really appreciated.

Thank you!

**Comment by halbmy** [comment]
I can't see why the original harmfit is apparently not working for your x and y positions. It is outdated (not pg>=1.1 style anyway). Maybe just an example where non-linear inversion of inverse problems fails with wrong starting model or regularization.

If you use the native implementation
```
from pygimli.frameworks.harmfit import harmfitNative as hf
```
all works well. 

![image](https://user-images.githubusercontent.com/7136401/186929096-b0d6ab47-667b-43e0-ab18-b831138a9bf8.png)

**Comment by halbmy** [comment]
`harmfit` is an inversion, and as in every inversion a meaningful error is needed. If any meaningful value (some cm, dm or m) is provided (all values below 10m seem to work)
```
x = hf(xx,tt,nc=10, error=1)[0]
```
all works similarly well. If not, a relative error of 1% (370m) is taken and then the result is random.

**Reply by benbpete** [reply]
This worked well.  Looks like I was just missing the error. Thanks again for the help!
